Confusing exception "ClassCastException: cannot assign instance of java.lang.invoke.SerializedLambda to field..."

If this code is on the server's classpath, the job executes successfully.

```
	public static DAG build() {
		DAG dag = new DAG();

        Vertex source = dag.newVertex("readMapP", readMapP("source"));
        // The "noop" vertex triggers the exception
        Vertex noop = dag.newVertex("mapP", mapP(entry -> entry));
        Vertex sink = dag.newVertex("writeLoggerP", writeLoggerP());

        dag.edge(Edge.between(source, noop));
        dag.edge(Edge.between(noop, sink));

		return dag;
	}
```

If the job is launched from the Java client, and the code is on the client side classpath but not the serverside, you get this exception:

```
Caused by: java.lang.ClassCastException: cannot assign instance of
java.lang.invoke.SerializedLambda to field 
com.hazelcast.jet.core.ProcessorMetaSupplier$1.val$addressToSupplier of type 
com.hazelcast.jet.function.DistributedFunction in instance of 
com.hazelcast.jet.core.ProcessorMetaSupplier$1
```

It's the middle vertex, _noop_, which causes the fault. Remove it and it's fine from client or server submission.

Tested on 0.5.1 and 0.7-SNAPSHOT (today, 2018-09-13).

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/JobConfig.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/JobConfig.java
@@ -249,6 +249,8 @@ public class JobConfig implements Serializable {
      * Adds the supplied classes to the list of resources that will be
      * available on the job's classpath while it's executing in the Jet
      * cluster.
+     * <p>
+     * See also {@link #addJar} and {@link #addResource}.
      *
      * @return {@code this} instance for fluent API
      */

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/CustomClassLoadedObject.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/CustomClassLoadedObject.java
@@ -17,6 +17,7 @@
 package com.hazelcast.jet.impl.execution.init;
 
 import com.hazelcast.jet.impl.serialization.SerializerHookConstants;
+import com.hazelcast.jet.impl.util.ExceptionUtil;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.Data;
@@ -62,6 +63,8 @@ public final class CustomClassLoadedObject {
         Thread.currentThread().setContextClassLoader(cl);
         try {
             return serializationService.toObject(data);
+        } catch (HazelcastSerializationException e) {
+            throw ExceptionUtil.handleSerializedLambdaCce(e);
         } finally {
             Thread.currentThread().setContextClassLoader(previous);
         }

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/ExceptionUtil.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/ExceptionUtil.java
@@ -30,6 +30,7 @@ import com.hazelcast.jet.datamodel.Tuple3;
 import com.hazelcast.jet.impl.exception.EnteringPassiveClusterStateException;
 import com.hazelcast.jet.impl.exception.ShutdownInProgressException;
 import com.hazelcast.logging.ILogger;
+import com.hazelcast.nio.serialization.HazelcastSerializationException;
 import com.hazelcast.spi.exception.CallerNotMemberException;
 import com.hazelcast.spi.exception.TargetNotMemberException;
 
@@ -180,5 +181,30 @@ public final class ExceptionUtil {
     @Nonnull
     public static <T extends Throwable> RuntimeException sneakyThrow(@Nonnull Throwable t) throws T {
         throw (T) t;
+    }
+
+    /**
+     * If the given exception has "java.lang.ClassCastException: cannot assign
+     * instance of java.lang.invoke.SerializedLambda" in the causes, wrap it in
+     * another JetException explaining the possible reason.
+     * <p>
+     * This is a hack to improve readability of this common exception.
+     *
+     * @param e the exception to handle
+     * @return the given exception wrapped, if it is a case of CCE for SerializedLambda
+     *     or the given exception otherwise
+     */
+    public static RuntimeException handleSerializedLambdaCce(HazelcastSerializationException e) {
+        Throwable cause = e.getCause();
+        while (cause != null) {
+            if (cause instanceof ClassCastException
+                    && cause.getMessage().startsWith("cannot assign instance of java.lang.invoke.SerializedLambda")) {
+                throw new JetException("Class containing the lambda probably missing from class path, did you add it " +
+                        "using JobConfig.addClass()?: " + e, e);
+            }
+            cause = cause.getCause();
+        }
+
+        throw e;
     }
 }

